File: src/libs/rust/m3/src/session/pager.rs
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use bitflags::bitflags;
use core::fmt;

use crate::cap;
use crate::com::{MemGate, RGateArgs, RecvGate, SendGate};
use crate::errors::Error;
use crate::goff;
use crate::int_enum;
use crate::kif;
use crate::session::ClientSession;
use crate::syscalls;
use crate::tiles::ChildActivity;

/// Represents a session at the pager.
///
/// The pager is responsible to resolve page faults and allows to create memory mappings.
pub struct Pager {
    sess: ClientSession,
    req_sgate: SendGate,
    child_sgate: cap::Selector,
    pf_rgate: RecvGate,
    pf_sgate: SendGate,
    close: bool,
}

int_enum! {
    /// The pager's operations
    pub struct PagerOp : u32 {
        /// A page fault
        const PAGEFAULT = 0x0;
        /// Initializes the pager session
        const INIT      = 0x1;
        /// Adds a child activity to the pager session
        const ADD_CHILD = 0x2;
        /// Adds a new send gate to the pager session
        const ADD_SGATE = 0x3;
        /// Clone the address space of a child activity (see `ADD_CHILD`) from the parent
        const CLONE     = 0x4;
        /// Add a new mapping with anonymous memory
        const MAP_ANON  = 0x5;
        /// Add a new data space mapping (e.g., a file)
        const MAP_DS    = 0x6;
        /// Add a new mapping for a given memory capability
        const MAP_MEM   = 0x7;
        /// Remove an existing mapping
        const UNMAP     = 0x8;
        /// Close the pager session
        const CLOSE     = 0x9;
    }
}

bitflags! {
    /// The mapping flags
    pub struct MapFlags : u32 {
        /// A private mapping, not shared with anyone else
        const PRIVATE = 0x0;
        /// A shared mapping
        const SHARED  = 0x2000;
        /// Do not initialize the memory
        const UNINIT  = 0x4000;
        /// Do not use a large page, even if possible
        const NOLPAGE = 0x8000;
    }
}

impl Pager {
    fn get_sgate(sess: &ClientSession) -> Result<cap::Selector, Error> {
        sess.obtain(
            1,
            |os| os.push_word(u64::from(PagerOp::ADD_SGATE.val)),
            |_| Ok(()),
        )
        .map(|crd| crd.start())
    }

    /// Creates a new session with given `SendGate` (for the pager).
    pub fn new(
        sess: ClientSession,
        pf_sgate: SendGate,
        child_sgate: cap::Selector,
    ) -> Result<Self, Error> {
        let pf_rgate = RecvGate::new_with(RGateArgs::default().order(6).msg_order(6))?;

        Ok(Pager {
            sess,
            req_sgate: SendGate::new_bind(kif::INVALID_SEL),
            child_sgate,
            pf_rgate,
            pf_sgate,
            close: false,
        })
    }

    /// Binds a new pager-session to given selector (for childs).
    #[cfg(not(target_vendor = "host"))]
    pub(crate) fn new_bind(sess_sel: cap::Selector, sgate_sel: cap::Selector) -> Self {
        let sess = ClientSession::new_bind(sess_sel);
        let sgate = SendGate::new_bind(sgate_sel);
        Pager {
            sess,
            req_sgate: sgate,
            child_sgate: kif::INVALID_SEL,
            pf_rgate: RecvGate::new_bind(kif::INVALID_SEL, 6, 6),
            pf_sgate: SendGate::new_bind(kif::INVALID_SEL),
            close: false,
        }
    }

    /// Clones the session to be shared with the given activity.
    pub(crate) fn new_clone(&self) -> Result<Self, Error> {
        let res = self.sess.obtain(
            1,
            |os| os.push_word(u64::from(PagerOp::ADD_CHILD.val)),
            |_| Ok(()),
        )?;
        let sess = ClientSession::new_bind(res.start());

        // get send gates for us and our child
        let child_sgate = Self::get_sgate(&sess)?;
        let req_sgate = SendGate::new_bind(Self::get_sgate(&sess)?);
        let pf_sgate = SendGate::new_bind(Self::get_sgate(&sess)?);

        let pf_rgate = RecvGate::new_with(RGateArgs::default().order(6).msg_order(6))?;
        Ok(Pager {
            sess,
            child_sgate,
            req_sgate,
            pf_rgate,
            pf_sgate,
            close: true,
        })
    }

    /// Initializes this pager session by delegating the activity cap to the server.
    pub(crate) fn init(&mut self, act: &ChildActivity) -> Result<(), Error> {
        // activate send and receive gate for page faults
        syscalls::activate(act.sel() + 1, self.pf_sgate.sel(), kif::INVALID_SEL, 0)?;
        syscalls::activate(act.sel() + 2, self.pf_rgate.sel(), kif::INVALID_SEL, 0)?;

        // delegate session and sgate caps to child
        act.delegate_obj(self.sel())?;
        act.delegate_obj(self.sgate_sel())?;

        // we only need to do that for clones
        if self.close {
            let crd = kif::CapRngDesc::new(kif::CapType::OBJECT, act.sel(), 1);
            self.sess.delegate(
                crd,
                |os| os.push_word(u64::from(PagerOp::INIT.val)),
                |_| Ok(()),
            )
        }
        else {
            Ok(())
        }
    }

    /// Returns the sessions capability selector.
    pub fn sel(&self) -> cap::Selector {
        self.sess.sel()
    }

    /// Returns the send gate capability selector for the child.
    pub(crate) fn sgate_sel(&self) -> cap::Selector {
        self.child_sgate
    }

    /// Performs the clone-operation on server-side using copy-on-write.
    #[allow(clippy::should_implement_trait)]
    pub fn clone(&self) -> Result<(), Error> {
        send_recv_res!(&self.req_sgate, RecvGate::def(), PagerOp::CLONE).map(|_| ())
    }

    /// Sends a page fault for the virtual address `addr` for given access type to the server.
    pub fn pagefault(&self, addr: goff, access: u32) -> Result<(), Error> {
        send_recv_res!(
            &self.req_sgate,
            RecvGate::def(),
            PagerOp::PAGEFAULT,
            addr,
            access
        )
        .map(|_| ())
    }

    /// Maps `len` bytes of anonymous memory to virtual address `addr` with permissions `prot`.
    pub fn map_anon(
        &self,
        addr: goff,
        len: usize,
        prot: kif::Perm,
        flags: MapFlags,
    ) -> Result<goff, Error> {
        let mut reply = send_recv_res!(
            &self.req_sgate,
            RecvGate::def(),
            PagerOp::MAP_ANON,
            addr,
            len,
            prot.bits(),
            flags.bits()
        )?;
        reply.pop()
    }

    /// Maps a dataspace of `len` bytes handled by given session to virtual address `addr` with
    /// permissions `prot`.
    pub fn map_ds(
        &self,
        addr: goff,
        len: usize,
        off: usize,
        prot: kif::Perm,
        flags: MapFlags,
        sess: &ClientSession,
    ) -> Result<goff, Error> {
        let crd = kif::CapRngDesc::new(kif::CapType::OBJECT, sess.sel(), 1);
        let mut res = 0;
        self.sess.delegate(
            crd,
            |os| {
                os.push_word(u64::from(PagerOp::MAP_DS.val));
                os.push_word(addr as u64);
                os.push_word(len as u64);
                os.push_word(u64::from(prot.bits()));
                os.push_word(u64::from(flags.bits()));
                os.push_word(off as u64);
            },
            |is| {
                res = is.pop_word()? as goff;
                Ok(())
            },
        )?;

        Ok(res)
    }

    pub fn map_mem(
        &self,
        addr: goff,
        mem: &MemGate,
        len: usize,
        prot: kif::Perm,
    ) -> Result<goff, Error> {
        let crd = kif::CapRngDesc::new(kif::CapType::OBJECT, mem.sel(), 1);
        let mut res = 0;
        self.sess.delegate(
            crd,
            |os| {
                os.push_word(u64::from(PagerOp::MAP_MEM.val));
                os.push_word(addr as u64);
                os.push_word(len as u64);
                os.push_word(u64::from(prot.bits()));
            },
            |is| {
                res = is.pop_word()? as goff;
                Ok(())
            },
        )?;

        Ok(res)
    }

    /// Unaps the mapping at virtual address `addr`.
    pub fn unmap(&self, addr: goff) -> Result<(), Error> {
        send_recv_res!(&self.req_sgate, RecvGate::def(), PagerOp::UNMAP, addr).map(|_| ())
    }
}

impl Drop for Pager {
    fn drop(&mut self) {
        if self.close {
            send_recv_res!(&self.req_sgate, RecvGate::def(), PagerOp::CLOSE).ok();
        }
    }
}

impl fmt::Debug for Pager {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
        write!(f, "Pager[sel: {}]", self.sel(),)
    }
}

File: src/libs/m3/session/Pager.cc
/*
 * Copyright (C) 2016-2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <m3/Syscalls.h>
#include <m3/com/GateStream.h>
#include <m3/com/OpCodes.h>
#include <m3/session/Pager.h>
#include <m3/tiles/ChildActivity.h>

namespace m3 {

Pager::Pager(capsel_t sess)
    : RefCounted(),
      ClientSession(sess, 0),
      _req_sgate(connect()),
      _child_sgate(connect().sel()),
      _pf_rgate(RecvGate::create(nextlog2<64>::val, nextlog2<64>::val)),
      _pf_sgate(connect()) {
}

Pager::Pager(capsel_t sess, capsel_t sgate)
    : RefCounted(),
      ClientSession(sess),
      _req_sgate(SendGate::bind(sgate)),
      _child_sgate(ObjCap::INVALID),
      _pf_rgate(RecvGate::bind(ObjCap::INVALID)),
      _pf_sgate(SendGate::bind(ObjCap::INVALID)) {
}

void Pager::pagefault(goff_t addr, uint access) {
    GateIStream reply = send_receive_vmsg(_req_sgate, opcodes::Pager::PAGEFAULT, addr, access);
    reply.pull_result();
}

void Pager::map_anon(goff_t *virt, size_t len, int prot, int flags) {
    GateIStream reply =
        send_receive_vmsg(_req_sgate, opcodes::Pager::MAP_ANON, *virt, len, prot, flags);
    reply.pull_result();
    reply >> *virt;
}

void Pager::map_ds(goff_t *virt, size_t len, int prot, int flags, const ClientSession &sess,
                   size_t offset) {
    KIF::ExchangeArgs args;
    ExchangeOStream os(args);
    os << opcodes::Pager::MAP_DS << *virt << len << prot << flags << offset;
    args.bytes = os.total();

    delegate(KIF::CapRngDesc(KIF::CapRngDesc::OBJ, sess.sel()), &args);

    ExchangeIStream is(args);
    is >> *virt;
}

void Pager::map_mem(goff_t *virt, MemGate &mem, size_t len, int prot) {
    KIF::ExchangeArgs args;
    ExchangeOStream os(args);
    os << opcodes::Pager::MAP_MEM << *virt << len << prot;
    args.bytes = os.total();

    delegate(KIF::CapRngDesc(KIF::CapRngDesc::OBJ, mem.sel()), &args);

    ExchangeIStream is(args);
    is >> *virt;
}

void Pager::unmap(goff_t virt) {
    GateIStream reply = send_receive_vmsg(_req_sgate, opcodes::Pager::UNMAP, virt);
    reply.pull_result();
}

Reference<Pager> Pager::create_clone() {
    KIF::CapRngDesc caps;
    {
        KIF::ExchangeArgs args;
        ExchangeOStream os(args);
        os << opcodes::Pager::ADD_CHILD;
        args.bytes = os.total();
        caps = obtain(1, &args);
    }

    return Reference<Pager>(new Pager(caps.start()));
}

void Pager::init(ChildActivity &act) {
    // activate send and receive gate for page faults
    Syscalls::activate(act.sel() + 1, _pf_sgate.sel(), KIF::INV_SEL, 0);
    Syscalls::activate(act.sel() + 2, _pf_rgate.sel(), KIF::INV_SEL, 0);

    // delegate the session cap
    act.delegate_obj(sel());
    // delegate request send gate for child
    act.delegate_obj(_child_sgate);

    // we only need to do that for clones
    if(!(flags() & KEEP_CAP)) {
        KIF::ExchangeArgs args;
        ExchangeOStream os(args);
        os << opcodes::Pager::INIT;
        args.bytes = os.total();
        delegate(KIF::CapRngDesc(KIF::CapRngDesc::OBJ, act.sel()), &args);
    }
}

void Pager::clone() {
    GateIStream reply = send_receive_vmsg(_req_sgate, opcodes::Pager::CLONE);
    reply.pull_result();
}

}
